### PR TITLE
chore(preview-env): New composite action for cleaning preview environments

### PR DESCRIPTION
**Pull request by clementnero**
Related to https://github.com/camunda/team-infrastructure/issues/300

Add a new composite action `preview-env/clean` for cleaning up preview environments:

- Support multiple preview labels (e.g. `deploy-preview` and `deploy-preview-ldap` for Cawemo).
- Lifetime of preview environment(s) associated to a PR is calculated from the date of the last deployment activity, if at least one environment is active.
- To reset lifetime, user has to trigger a new deployment (e.g. push new commits, re-run deploy workflow).
- If lifetime reaches `ttl` threshold, associated PR is unlabeled (triggering the cleanup cycle) and a comment is added to the PR.
- If lifetime reaches `warning-ttl` threshold (and is lower than `ttl`), a warning comment is added to the PR.
- Great care is taken to inform users of the status of preview environment(s). Comments are dynamically added, edited or deleted depending on the status, and contain exhaustive information.
- Thresholds and messages are customizable.

Multiple tests have been made on all repositories that support preview environments. To facilitate testing, a temporary commit has been pushed to put the deploy and teardown workflows (ArgoCD steps) into "dry-run" mode, and to restrict the process to my related PRs (to not affect existing preview environments).

> The reviewer can test this new composite action directly from the related PRs (e.g. https://github.com/camunda/cawemo/pull/11342).

Before merging:
- [ ] Remove temporary commit **do-not-merge(preview-env): Changes for testing purposes (dry-run) -> disable Argocd steps + process only gh-infra-300 feature branch**

### CODE REVIEW

**Review by Langleu — COMMENTED** [review]
would recommend moving the bash scripts to separate files as well.
Would help with readability, syntax highlighting, linting, etc.

File: preview-env/clean/graphql/query_get_label_id.gql [hunk 1]
```diff
@@ -0,0 +1,9 @@
+query get_label_id ($owner:String!,$repo:String!,$label:String!) {
+
+  repository (name: $repo, owner: $owner) {
+    label (name: $label) {
+      id
+    }
+  }
+  
+}
```
Comment (Langleu) [inline preview-env/clean/graphql/query_get_label_id.gql:9]: newline missing, just for completeness.
Would recommend adding pre-commit locally `pre-commit install` to run on this repo.

File: preview-env/clean/action.yml [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+          jq \
+            --arg unique_comment_tag "$unique_comment_tag" \
+            -r \
+            '(.[] | select(.body | contains($unique_comment_tag))).id'
+        )
+
+        # Comment content to be created or updated if necessary
+        new_comment_body=""
+
+        echo "Processing PR #$pr_id (lifetime=${lifetime}s,ttl=${TTL}s,warning-ttl=${WARNING_TTL}s)"
```
Comment (Langleu) [inline preview-env/clean/action.yml:273]: any chance we could use the PR number instead?
`#PR_kwDOA5eBQs5YH_uf` vs `#1342` would help with debugging.
The id is part of the source code of a page but if you have 10+ preview-envs it will be easier to use the PR number for debugging / plausibility checks.

File: preview-env/clean/action.yml [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+          --field owner=$REPOSITORY_OWNER \
+          --field repo=$REPOSITORY_NAME \
+          --field 'states[]=CLOSED' \
+          --field 'states[]=MERGED' \
+          --field query=@${{ github.action_path }}/graphql/query_get_pull_requests_with_comments.gql \
+          --jq '.data.repository.pullRequests.nodes'
+      )
+
+      # If Pull Request is not OPEN, any warning or shutdown comments must be removed
+      # If Pull Request is OPEN and has no deploy labels, any warning comments must be removed
```
Comment (Langleu) [inline preview-env/clean/action.yml:403]: maybe rephrase the comment

I assume it's supposed to delete the "delete in x days" comment but keep the "deleted env" comment?

Both include a "warning" in the message itself, which can be misleading.

**Comment by clementnero** [comment]
> would recommend moving the bash scripts to separate files as well. Would help with readability, syntax highlighting, linting, etc.

Good point, I will move part of the bash scripts to separate files.

File: preview-env/clean/graphql/query_get_label_id.gql (hunk 1 above)
Comment (clementnero) [reply preview-env/clean/graphql/query_get_label_id.gql:9]: Good point.

File: preview-env/clean/action.yml (hunk 2 above)
Comment (clementnero) [reply preview-env/clean/action.yml:273]: Good point. I will switch to PR number.

File: preview-env/clean/action.yml (hunk 3 above)
Comment (clementnero) [reply preview-env/clean/action.yml:403]: For the shutdown message, I suggest to replace the warning block with a note block to avoid confusion.
image

> I assume it's supposed to delete the "delete in x days" comment but keep the "deleted env" comment?

Yes only for OPEN PRs with no preview labels attached.

Comment (Langleu) [reply preview-env/clean/action.yml:403]: I got confused by it in the first place as this has `warning` as part of the message, so I thought it may delete it but doesn't since it's not the `warning` message that is meant.
image

Comment (clementnero) [reply preview-env/clean/action.yml:403]: Let's switch to a note block 👍

**Comment by clementnero** [comment]
> 

The scripts are now more verbose as suggested.

**Review by Langleu — APPROVED** [review]
don't forget to run pre-commit to fix the trailing whitespace.
Inspiring work done, detailed and well executed!
I like the verbosity that you added and thanks for outsourcing the scripts to separate files, helps with the readability and allows to run shellcheck more easily.
